src/libmw/include/mw/mmr/MMRUtil.h

```diff
@@ -0,0 +1,147 @@
+        } else {
+            uint64_t increment = m_baseInc + BitUtil::CountRightmostZeros(m_siblingNum);
+            m_next = mmr::Index(m_next.GetPosition() + increment, m_height);
+        }
+
+        ++m_siblingNum;
+        return m_next <= m_lastNode;
+    }
+
+    const mmr::Index& Get() const noexcept { return m_next; }
+    const uint64_t GetPosition() const noexcept { return m_next.GetPosition(); }
+
+private:
+    uint64_t m_height;
+    mmr::Index m_lastNode;
+    uint64_t m_baseInc;
+
+    uint64_t m_siblingNum;
+    mmr::Index m_next;
+};
+
+class MMRUtil
+{
+public:
+    static BitSet BuildCompactBitSet(const uint64_t num_leaves, const BitSet& unspent_leaf_indices)
+    {
+        BitSet compactable_node_indices(num_leaves * 2);
+
+        boost::dynamic_bitset<> prunable_nodes(num_leaves * 2);
+
+        mmr::LeafIndex leaf_idx = mmr::LeafIndex::At(0);
+        while (leaf_idx.GetLeafIndex() < num_leaves) {
+            if (unspent_leaf_indices.size() > leaf_idx.GetLeafIndex() && !unspent_leaf_indices.test(leaf_idx.GetLeafIndex())) {
+                prunable_nodes.set(leaf_idx.GetPosition());
+            }
+
+            leaf_idx = leaf_idx.Next();
+        }
+
+        mmr::Index last_node = mmr::LeafIndex::At(num_leaves).GetNodeIndex();
+
+        uint64_t height = 1;
+        while ((std::pow(2, height + 1) - 2) <= last_node.GetPosition()) {
+            SiblingIter iter(height, last_node);
+            while (iter.Next()) {
+                mmr::Index right_child = iter.Get().GetRightChild();
+                if (prunable_nodes.test(right_child.GetPosition())) {
+                    mmr::Index left_child = iter.Get().GetLeftChild();
+                    if (prunable_nodes.test(left_child.GetPosition())) {
+                        compactable_node_indices.set(right_child.GetPosition());
+                        compactable_node_indices.set(left_child.GetPosition());
+                        prunable_nodes.set(iter.Get().GetPosition());
+                    }
+                }
+            }
+
+            ++height;
+        }
+
+        return compactable_node_indices;
+    }
+
+    static BitSet DiffCompactBitSet(const BitSet& prev_compact, const BitSet& new_compact)
+    {
+        BitSet diff;
+
+        for (size_t i = 0; i < new_compact.size(); i++) {
+            if (prev_compact.size() > i && prev_compact.test(i)) {
+                assert(new_compact.test(i));
+                continue;
+            }
+
+            diff.bitset.push_back(new_compact.test(i));
+        }
+
+        return diff;
+    }
+
+    /// <summary>
+    /// An MMR can be rebuilt from the leaves and a small set of carefully chosen parent hashes.
+    /// This calculates the positions of those parent hashes.
+    /// </summary>
+    /// <param name="unspent_leaf_indices">The unspent leaf indices.</param>
+    /// <returns>The pruned parent positions.</returns>
+    static BitSet CalcPrunedParents(const BitSet& unspent_leaf_indices)
```

Comment:
This function is pretty much identical to BuildCompactBitSet, can we just have one function please?

Reply:
It's not quite the same, but I'm sure I can find a way to combine them.